source/core/slang-uint-set.cpp

```diff
@@ -162,5 +143,24 @@ void UIntSet::intersectWith(const UIntSet& set)
     return false;
 }
 
+Index UIntSet::countElements() const
+{
+    // TODO: This can be made faster using SIMD intrinsics to count set bits.
+    uint64_t tmp;
+    constexpr Index loopSize = ((sizeof(Element) / sizeof(tmp)) != 0) ? sizeof(Element) / sizeof(tmp) : 1;
+    Index count = 0;
+    for (auto index = 0; index < this->m_buffer.getCount(); index++)
+    {
+        for (auto i = 0; i < loopSize; i++)
+        {
+            tmp = m_buffer[index] >> (sizeof(tmp) * i);
+            tmp = tmp - ((tmp >> 1) & 0x5555555555555555);
+            tmp = (tmp & 0x3333333333333333) + ((tmp >> 2) & 0x3333333333333333);
+            count += ((tmp + (tmp >> 4) & 0xF0F0F0F0F0F0F0F) * 0x101010101010101) >> 56;
+        }
+    }
+    return count;
+}
```

Comment:
Since this looks to be performance related, it would be better to optimize more.
1. Indexing `m_buffer` for every loop seems a waste.
2. shifting with a runtime variable is not good for optimization.
3. `this->m_buffer.getCount()` could be inlined but may not; it doesn't need to dereference `this` all the time as  well.
4. when the 64bit whole block is filled with zero, we can simply skip the block.

```
Index UIntSet::countElements() const
{
    // TODO: This can be made faster using SIMD intrinsics to count set bits.
    Index count = 0;

    uint64_t tmp;
    Index loopSize = std::max<size_t>(1, sizeof(Element) / sizeof(tmp));
    Index bufferCount = m_buffer.getCount();
    for (Index index = 0; index < bufferCount; index++)
    {
        Element buf = m_buffer[index];
        for (Index i = 0; i < loopSize; i++)
        {
            tmp = static_cast<decltype(tmp)>(buf);
            buf = buf >> sizeof(tmp);
            if (tmp)
            {
                tmp = tmp - ((tmp >> 1) & 0x5555555555555555);
                tmp = (tmp & 0x3333333333333333) + ((tmp >> 2) & 0x3333333333333333);
                count += ((tmp + (tmp >> 4) & 0xF0F0F0F0F0F0F0F) * 0x101010101010101) >> 56;
            }
        }
    }
    return count;
}
```

Reply:
Out of curiosity I setup to profile the code

They perform the same +-100ns depending which run first and how the cache was used before the function was launched.

Reply:
I don't get why we needed the second layer loop though, can't we just accumulate with `Element` type instead of hard code it to be at uint64? For most targets, `Elment` is a 64bit integer so that inner loop is trivially 1 iteration anyways.

Reply:
I expect the second loop to be removed by the compiler since the loopSize is known at the compile time.

And yes, my code wouldn't make a much difference in practice.
But I wanted to share some experience for "how" to do the code optimization.

Note also that Slang could run on CPU other than Intel CPUs which may show different result.

Reply:
> I don't get why we needed the second layer loop though, can't we just accumulate with Element type instead of hard code it to be at uint64? For most targets, Elment is a 64bit integer so that inner loop is trivially 1 iteration anyways.

Swapping the internal data structure `Element` should break-minimal things (hopefully nothing), else it is implemented wrong.

Reply:
For the sake of argument, the second loop could be helpful.
I just remembered something and sharing it here for your information.

The CPU on PS4 doesn't support the 64 bit integer operations natively. It is emulated with two 32 bit instructions.
If that is the case, it could have a better chance to skip the while zeros in `tmp` when the size of tmo is smaller.